### ISSUE TITLE
Different DOFs at the two sides of shells or beams

### ISSUE THREAD

**Issue by qaumann** [KratosMultiphysics/Kratos#6738]
Currently I am implementing the coupling between a structure and a surrounding acoustic fluid. I was wondering, if it is possible, to use shells or beams in this context. The concept is described here:

![image](https://user-images.githubusercontent.com/27352976/78771504-89a89600-7990-11ea-8dfe-a84f9273b79d.png)

The problem here is, that the pressure in the two fluid domains on the opposite sides is not equal (p1 and p2 in the image), but I cannot define two dofs with the same key in a single node. Is there already something available in Kratos, which could help me? I think the problem also arises in FSI applications where shells are used.

I came across the negative and positive face pressure; is this how the problem is solved (i.e. mapping the surrounding pressure to either negative or positive face pressure, depending on the orientation of the shell)? Are there any examples on how to use this?

**Comment by rubenzorrilla** [comment]
What we normally do in the FSI analysis of membrane-like structures is to duplicate the geometry in the volume domain (i.e. the fluid). Hence, one side is the "positive" side, whose pressure will be mapped to the `POSITIVE_FACE_PRESSURE` variable the other side is the "negative" side, whose pressure will be mapped to the `NEGATIVE_FACE_PRESSURE`.
Once both pressure values are mapped to the structure, the load condition does the summation of both values to obtain the final load to be applied to the thin-walled body.

I've pretty much experience doing this and, aside of the problems that may arise when meshing duplicated curved surfaces (the polygonals of approximating both overlapped surfaces tend to intersect in the machine precision limit), it works pretty well. 

The real challenge is in the beams mapping, which as far as I know we have not yet resolved... (@philbucher can probably shed some light here).

**Comment by loumalouomega** [comment]
> What we normally do in the FSI analysis of membrane-like structures is to duplicate the geometry in the volume domain (i.e. the fluid). Hence, one side is the "positive" side, whose pressure will be mapped to the `POSITIVE_FACE_PRESSURE` variable the other side is the "negative" side, whose pressure will be mapped to the `NEGATIVE_FACE_PRESSURE`.
> Once both pressure values are mapped to the structure, the load condition does the summation of both values to obtain the final load to be applied to the thin-walled body.
> 
> I've pretty much experience doing this and, aside of the problems that may arise when meshing duplicated curved surfaces (the polygonals of approximating both overlapped surfaces tend to intersect in the machine precision limit), it works pretty well.
> 
> The real challenge is in the beams mapping, which as far as I know we have not yet resolved... (@philbucher can probably shed some light here).

Basically duplicate the geometries as @rubenzorrilla. If the ' POSITIVE_FACE_PRESSURE`/`NEGATIVE_FACE_PRESSURE` duo can be avoided if one of the geometry is flipped, then the normal is inverted and `PRESSURE` can be considered in both sides

**Comment by loumalouomega** [comment]
> > What we normally do in the FSI analysis of membrane-like structures is to duplicate the geometry in the volume domain (i.e. the fluid). Hence, one side is the "positive" side, whose pressure will be mapped to the `POSITIVE_FACE_PRESSURE` variable the other side is the "negative" side, whose pressure will be mapped to the `NEGATIVE_FACE_PRESSURE`.
> > Once both pressure values are mapped to the structure, the load condition does the summation of both values to obtain the final load to be applied to the thin-walled body.
> > I've pretty much experience doing this and, aside of the problems that may arise when meshing duplicated curved surfaces (the polygonals of approximating both overlapped surfaces tend to intersect in the machine precision limit), it works pretty well.
> > The real challenge is in the beams mapping, which as far as I know we have not yet resolved... (@philbucher can probably shed some light here).
> 

Basically duplicate the geometries as @rubenzorrilla. If the `POSITIVE_FACE_PRESSURE` `NEGATIVE_FACE_PRESSURE` duo can be avoided if one of the geometry is flipped, then the normal is inverted and `PRESSURE` can be considered in both sides, but then if `PRESSURE` is considered as DoF it will be duplicated. 

Another approach could be to duplicate the nodes and use constraints in order to impose the same displacement

**Comment by qaumann** [comment]
Thank you both very much for the input. I will definately try to duplicate the interface and I am very glad, that the workflow with negative and positive face pressure already exists. Do I have to use two models for the fluid domains or can this be done in one model using model parts? Is there perhaps a small test example available which shows how this is applied?

I would really like to try also @loumalouomega's approach, but I am not sure if I got it right. In this setup I would e.g. have some interface nodes for my left domain and some duplicated interface nodes for the right domain, which are also used for the shell. Now in order to compute the effect of the right domain on the shell, my coupling condition would not only need the interface nodes of the left domain, but also of the right domain, in order to access the shell's displacement dofs. Is this correct or am I missing something?

**Comment by rubenzorrilla** [comment]
> Thank you both very much for the input. I will definately try to duplicate the interface and I am very glad, that the workflow with negative and positive face pressure already exists. Do I have to use two models for the fluid domains or can this be done in one model using model parts? Is there perhaps a small test example available which shows how this is applied?
> 
I think that it can be achieved in several ways. What I do in the FSI is to create a unique model, with a unique volume of fluid, but i have two model parts for the skin (one for the positive side and one for the negative counterpart).

The challenge is in the preprocess. Taking a 3D case as example, imagine that your double-sided interface is a square. Such square has 4 lines as boundaries, these four lines are **unique** (not duplicated). From these 4 lines you need to create two **duplicated** surfaces, one overlapping each other.

I hope this explains (unfortunately I've no examples in the laptop...).

**Comment by qaumann** [comment]
Thanks for the explanation. As I understand it now, I double my interface (element and conditions) and define one as positive and one as negative side. Then I compute the load on my interface based on the mapped pressure and add resp. subtract positive and negative face pressure to my overall pressure (here I can use a standard surface load condition, right?). And it does not matter, that the pressure dof is identical, because it is mapped to the two different variables and handled differently in the condition. Is this correct?

This also means, that I do not need a special "coupling condition", but can use the standard load conditions, as long as my mapping is correct.

Regarding an example, I found the `non_conformant_one_side_map_test.py` which is quite illustrative in my oppinion.

**Comment by loumalouomega** [comment]
> Thanks for the explanation. As I understand it now, I double my interface (element and conditions) and define one as positive and one as negative side. Then I compute the load on my interface based on the mapped pressure and add resp. subtract positive and negative face pressure to my overall pressure (here I can use a standard surface load condition, right?). And it does not matter, that the pressure dof is identical, because it is mapped to the two different variables and handled differently in the condition. Is this correct?
> 
> This also means, that I do not need a special "coupling condition", but can use the standard load conditions, as long as my mapping is correct.
> 
> Regarding an example, I found the `non_conformant_one_side_map_test.py` which is quite illustrative in my oppinion.

If the dof is shared and if the values are independent it can be an issue

**Comment by qaumann** [comment]
The dof values should be independent in my oppinion; so I would need two separate dofs, one for each side of the interface. But then I should be able to use the mapping workflow, right? I would have to duplicate the nodes though, as I need the dofs separated.

**Comment by loumalouomega** [comment]
> The dof values should be independent in my oppinion; so I would need two separate dofs, one for each side of the interface. But then I should be able to use the mapping workflow, right? I would have to duplicate the nodes though, as I need the dofs separated.

That's the problem, if the dof   is pressure, which is linked to the normal, you should duplicate the nodes or find some tricky way as creating auxiliar dofs and link them with constraints

**Comment by rubenzorrilla** [comment]
I think it is much easier.

In the preprocess:
Fluid domain:
- The surface is duplicated (lets say `left_side` and `right_side` submodelparts).
Structure domain:
- You have a unique membrane / shell structure with `SurfaceLoadConditions` (lets call it  `SurfaceStructure` submodelpart).

When doing the coupling:
- From `left_side` to `SurfaceStructure`map `PRESSURE` to `POSITIVE_FACE_PRESSURE`.
- From `right_side` to `SurfaceStructure`map `PRESSURE`to `NEGATIVE_FACE_PRESSURE`.

Note that the normals in `left_side` and `right_side` need to be consistently oriented (both pointing inwards or outwards). Also note that this affects the sign of the load.

With regard to the mapping, my advise is to use the `MappingApplication`. The mappers in the `FSIApplication` are legacy.

**Comment by qaumann** [comment]
Thanks for summing it up for me. So the only thing needed to be duplicated is the interface surface (one `SurfaceLoadCondition` per side) but not the nodes or dofs, right?

I will try it using the `MappingApplication`, thanks for the hint.

**Comment by rubenzorrilla** [comment]
> Thanks for summing it up for me. So the only thing needed to be duplicated is the interface surface (one `SurfaceLoadCondition` per side) but not the nodes or dofs, right?
> 
Precisely. I insist that the duplicated surface is the fluid one (not the structure one).
